File: files/filedirectory/src/main/java/example/micronaut/filedirectory/FilesController.java
package example.micronaut.filedirectory;

import io.micronaut.http.MediaType;
import io.micronaut.http.annotation.Controller;
import io.micronaut.http.annotation.Get;
import io.micronaut.http.annotation.Produces;

import javax.validation.constraints.NotBlank;
import java.util.Optional;

@Controller("/files")
public class FilesController {

    @Produces(MediaType.TEXT_PLAIN)
    @Get("/stock/{isbn}")
    public Boolean stock(@NotBlank String isbn) {
        return fileDirectoryByIsbn(isbn).map(bi -> bi.getStock() > 0).orElse(null);
    }

    private Optional<FileDirectory> fileDirectoryByIsbn(String isbn) {
        if (isbn.equals("1491950358")) {
            return Optional.of(new FileDirectory(isbn, 4));

        } else if (isbn.equals("1680502395")) {
            return Optional.of(new FileDirectory(isbn, 0));
        }
        return Optional.empty();
    }
}

File: files/filedirectory/src/main/kotlin/example/micronaut/FilesController.kt
package example.micronaut

import io.micronaut.http.MediaType
import io.micronaut.http.annotation.Controller
import io.micronaut.http.annotation.Get
import io.micronaut.http.annotation.Produces
import java.util.Optional
import javax.validation.constraints.NotBlank

@Controller("/files") // <1>
open class FilesController {

    @Produces(MediaType.TEXT_PLAIN) // <2>
    @Get("/stock/{isbn}") // <3>
    open fun stock(@NotBlank isbn: String): Boolean? {
        return fileDirectoryByIsbn(isbn).map { (_, stock) -> stock > 0 }.orElse(null)
    }

    private fun fileDirectoryByIsbn(isbn: String): Optional<FileDirectory> {
        if (isbn == "1491950358") {
            return Optional.of(FileDirectory(isbn, 4))

        } else if (isbn == "1680502395") {
            return Optional.of(FileDirectory(isbn, 0))
        }
        return Optional.empty()
    }
}
